hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java

```diff
@@ -510,7 +511,23 @@ public OzoneVolume getVolumeDetails(String volumeName)
 
   @Override
   public S3VolumeContext getS3VolumeContext() throws IOException {
-    return ozoneManagerClient.getS3VolumeContext();
+    S3VolumeContext resp = ozoneManagerClient.getS3VolumeContext();
+    String userPrincipal = resp.getUserPrincipal();
+    updateS3Principal(userPrincipal);
+    return resp;
+  }
+
+  private void updateS3Principal(String userPrincipal) {
+    S3Auth s3Auth = this.getThreadLocalS3Auth();
+    // Update user principal if needed to be used for KMS client
+    if (s3Auth != null) {
+      // Update userPrincipal field with the value returned from OM. So that
+      //  in multi-tenancy, KMS client can use the correct identity
+      //  (instead of using accessId) to communicate with KMS.
+      LOG.debug("Updating S3Auth.userPrincipal to {}", userPrincipal);
+      s3Auth.setUserPrincipal(userPrincipal);
+      this.setThreadLocalS3Auth(s3Auth);
```

Comment:
hmm.. just for myself learning.. so does that mean a same user always access the same rpcClient instance so we use threadLocal variable? 
and is it possible a rpcClient instance used by different users?

Reply:
Basically, when a client calls S3G get api, the client's AwsAccessId is stored as the current UserPrincipal in threadlocal. Then S3G calls OM, through RpcClient, to grab the key information and OM internally resolves the user information to the right Kerberos  UserPrincipal and returns to S3G. RpcClient updates this actual Kerberos UserPrincipal in its thread context and uses it to talk to datanodes.

Reply:
got it! thanks Duong for the explanation!!